content/blog/2019-09-25-bril-type-inference.md

```diff
@@ -0,0 +1,204 @@
+of these variables are not ints, then we have a type unification error, and the
+program is not well-typed.
+
+## Implementation
+
+### Modifying the Grammar
+We want to make type annotations optional. So first I modified the grammar to
+have a rule for type annotations:
+```
+type_decl.5: ":" type
+const.4: IDENT [type_decl] "=" "const" lit ";"
+...
+```
+This is an issue however because there is an ambiguity between labels and
+assignments. Consider these two Bril programs:
+```
+main {
+l:
+  x = const 5;
+}
+```
+and
+```
+main {
+  l: x = const 5;
+}
+```
+These programs consist of the same tokens. However, the first one intuitively
+means that we have a label `l` and some int `x` with a value of 5. The second
+one means that we have a variable called `l` of type `x` with a value of 5.
+It is incorrect to simply have one rule as a higher priority, because
+semantically, both of these should be allowed as separate programs. So we have
+two options:
+1. Force labels to be on its own line
+2. Only allow fixed type names, like "int" and "bool"
+
+I decided to go with (2) because I didn't know how to easily do (1), but in
+retrospect (1) may have been better, because some other students' projects
+allow for user-defined types.
+
+### Type Inference
+As noted in the design, inferring the types of variables for a single statement
+is straightforward. Here is an example snippet showing how types for comparison
+ops are inferred:
+```
+...
+elif instr["op"] in COMPARISON_OPS:
+    for arg in instr["args"]:
+        type_var(gamma, arg, "int", i)
+    type_var(gamma, instr["dest"], "bool", i)
+...
+```
+Here, we keep track of a typing context `gamma` which maps variables to their
+type. Then, we check that each argument is either untyped or already has the
+type `int`; otherwise we'll throw a type unification error. Finally, we do
+the same for the destination, making sure that it is typed as `bool`.
+
+For our implementation, we simply iterate through each instruction and determine
+the types of our variables. The one case where this presents issues is with
+`id`. In general, `id` sets the type of the destination to the type of the
+variable on the right hand side. However, what happens if we don't know the type
+of the variable being copied? For example, consider this Bril program:
+```
+main {
+  jmp later;
+earlier:
+  x = id y;
+  ret;
+later:
+  y = const 5;
+  jmp earlier;
+}
+```
+This is a valid program that should typecheck. Specifically, `x` and `y` are
+both ints. However, if we naively go sequentially through each instruction, we
+don't know the type of `y` until we have `y = const 5`, at which point it is too
+late to type `x`. To resolve this, we have two options:
+1. Rerun the type inference algorithm, stopping when no more variables have been
+inferred
+2. Keep track of which variables must have the same type, and after 1 pass,
+setting their types to be the same.
```

Comment:
It's worth noting that this is basically constraint-based type inference (i.e., how type inference works for "real" languages).

Reply:
:O did not know this. TIL